File: entroly/work_context_snapshot_store.py
"""Hardened host-byte store for Work Graph context snapshots.

Semantic truth stays elsewhere:
- ContextReceipt/RecoveryHandle meaning is Rust-owned.
- Work Graph state stores only bounded receipt/handle references.
- This module stores the exact canonical host context JSON needed to service a
  later page fault without placing those source bytes inside the graph or MCP
  token.

A snapshot token is derived from the verified repository-context
``receipt.context_sha256``. The canonical Rust ContextReceipt already stores the
same value as ``source_commitment``; therefore another agent can reconstruct the
short snapshot locator from durable receipt evidence instead of depending on a
previous MCP response surviving verbatim.

The verified-context commitment intentionally excludes volatile host metadata
(``generation`` and ``command``). Those fields are stripped before persistence
as well, so one commitment maps to one stable byte representation rather than
multiple snapshots that merely differ in process-local metadata.

The store deliberately reuses ``WorkGraphStore``'s repository directory and
cross-process lock so context snapshots follow the same local isolation and
persistence boundary without inventing another coordination protocol.
"""
from __future__ import annotations

import copy
import json
import os
import stat
import tempfile
from pathlib import Path
from typing import Any

from .work_graph_store import WorkGraphStore, WorkGraphStateError, _fsync_dir, _private_dir

CONTEXT_SNAPSHOT_TOKEN_PREFIX = "wctx1."
DEFAULT_MAX_CONTEXT_BYTES = 512 * 1024
DEFAULT_MAX_SNAPSHOTS = 8_192
DEFAULT_MAX_TOTAL_BYTES = 256 * 1024 * 1024
_HEX = frozenset("0123456789abcdef")
_VOLATILE_CONTEXT_FIELDS = frozenset({"generation", "command"})


class WorkContextSnapshotError(WorkGraphStateError):
    """A context snapshot is malformed, unsafe, missing, or exceeds bounds."""


def _verify_snapshot_bytes(raw: bytes, expected_digest: str) -> None:
    """Delegate v1 snapshot commitment semantics to the Rust kernel."""
    try:
        from entroly_core import verified_context_snapshot_verify_bytes
    except (ImportError, AttributeError) as exc:
        raise WorkContextSnapshotError(
            "native context snapshot verifier is unavailable"
        ) from exc
    try:
        actual = verified_context_snapshot_verify_bytes(raw, expected_digest)
    except (TypeError, ValueError) as exc:
        raise WorkContextSnapshotError(str(exc)) from exc
    if actual != expected_digest:
        raise WorkContextSnapshotError(
            "native context snapshot verifier returned a different commitment"
        )


class WorkContextSnapshotStore:
    """Repository-scoped exact context bytes keyed by verified context identity."""

    def __init__(
        self,
        graph_store: WorkGraphStore,
        *,
        max_context_bytes: int = DEFAULT_MAX_CONTEXT_BYTES,
        max_snapshots: int = DEFAULT_MAX_SNAPSHOTS,
        max_total_bytes: int = DEFAULT_MAX_TOTAL_BYTES,
    ) -> None:
        if not isinstance(graph_store, WorkGraphStore):
            raise TypeError("graph_store must be a WorkGraphStore")
        if isinstance(max_context_bytes, bool) or int(max_context_bytes) < 1_024:
            raise ValueError("max_context_bytes must be an integer >= 1024")
        if isinstance(max_snapshots, bool) or not 1 <= int(max_snapshots) <= 100_000:
            raise ValueError("max_snapshots must be an integer between 1 and 100000")
        if isinstance(max_total_bytes, bool) or int(max_total_bytes) < int(max_context_bytes):
            raise ValueError("max_total_bytes must be at least max_context_bytes")
        self.graph_store = graph_store
        self.max_context_bytes = int(max_context_bytes)
        self.max_snapshots = int(max_snapshots)
        self.max_total_bytes = int(max_total_bytes)
        self.context_dir = graph_store.repo_dir / "context-snapshots"
        _private_dir(self.context_dir)

    @staticmethod
    def _valid_digest(value: object) -> bool:
        return (
            isinstance(value, str)
            and len(value) == 64
            and all(ch in _HEX for ch in value)
        )

    @classmethod
    def token_for_commitment(cls, commitment: str) -> str:
        """Derive the durable short locator carried by MCP/continuation state."""
        if not cls._valid_digest(commitment):
            raise WorkContextSnapshotError("context commitment is not a sha256 digest")
        return CONTEXT_SNAPSHOT_TOKEN_PREFIX + commitment

    @classmethod
    def _digest_from_token(cls, token: str) -> str:
        if not isinstance(token, str) or not token.startswith(CONTEXT_SNAPSHOT_TOKEN_PREFIX):
            raise WorkContextSnapshotError("unsupported context snapshot token")
        digest = token[len(CONTEXT_SNAPSHOT_TOKEN_PREFIX):]
        if not cls._valid_digest(digest):
            raise WorkContextSnapshotError("context snapshot token has an invalid digest")
        return digest

    @classmethod
    def _context_commitment(cls, payload: dict[str, Any]) -> str:
        receipt = payload.get("receipt")
        digest = receipt.get("context_sha256") if isinstance(receipt, dict) else None
        if not cls._valid_digest(digest):
            raise WorkContextSnapshotError(
                "context snapshot is missing a valid verified context commitment"
            )
        return str(digest)

    @staticmethod
    def _stable_payload(payload: dict[str, Any]) -> dict[str, Any]:
        """Return the exact semantic payload covered by ``context_sha256``.

        The verified-context commitment excludes only process-local
        ``generation``/``command`` fields. Persisting them under that commitment
        would violate content-address uniqueness, so storage follows the same
        scope exactly.
        """
        stable = copy.deepcopy(payload)
        for field in _VOLATILE_CONTEXT_FIELDS:
            stable.pop(field, None)
        return stable

    def _canonical_bytes(self, payload: dict[str, Any]) -> bytes:
        try:
            raw = json.dumps(
                payload,
                sort_keys=True,
                ensure_ascii=True,
                separators=(",", ":"),
                allow_nan=False,
            ).encode("utf-8")
        except (TypeError, ValueError) as exc:
            raise ValueError("context snapshot must be canonical JSON") from exc
        if len(raw) > self.max_context_bytes:
            raise WorkContextSnapshotError(
                f"context snapshot exceeds {self.max_context_bytes} bytes"
            )
        return raw

    def _path(self, digest: str) -> Path:
        return self.context_dir / f"{digest}.json"

    def _read_path_unlocked(self, path: Path, expected_digest: str) -> dict[str, Any]:
        try:
            info = path.lstat()
        except FileNotFoundError as exc:
            raise WorkContextSnapshotError("context snapshot is unavailable") from exc
        except OSError as exc:
            raise WorkContextSnapshotError(f"cannot inspect context snapshot: {exc}") from exc
        if stat.S_ISLNK(info.st_mode) or not stat.S_ISREG(info.st_mode):
            raise WorkContextSnapshotError("unsafe context snapshot path")
        if info.st_size > self.max_context_bytes:
            raise WorkContextSnapshotError("context snapshot exceeds its byte bound")
        flags = os.O_RDONLY | getattr(os, "O_NOFOLLOW", 0)
        try:
            fd = os.open(path, flags)
        except OSError as exc:
            raise WorkContextSnapshotError(f"cannot open context snapshot safely: {exc}") from exc
        try:
            current = os.fstat(fd)
            if not stat.S_ISREG(current.st_mode) or current.st_size > self.max_context_bytes:
                raise WorkContextSnapshotError("unsafe or oversized context snapshot")
            raw = bytearray()
            while len(raw) <= self.max_context_bytes:
                chunk = os.read(
                    fd,
                    min(1024 * 1024, self.max_context_bytes + 1 - len(raw)),
                )
                if not chunk:
                    break
                raw.extend(chunk)
        finally:
            os.close(fd)
        encoded = bytes(raw)
        if len(encoded) > self.max_context_bytes:
            raise WorkContextSnapshotError("context snapshot exceeds its byte bound")
        try:
            payload = json.loads(encoded.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            raise WorkContextSnapshotError("context snapshot is not valid JSON") from exc
        if not isinstance(payload, dict):
            raise WorkContextSnapshotError("context snapshot root must be an object")
        if any(field in payload for field in _VOLATILE_CONTEXT_FIELDS):
            raise WorkContextSnapshotError("context snapshot contains volatile host metadata")
        # The store writes one canonical byte representation. A same-semantics
        # whitespace/key-order rewrite is still a storage mutation and is
        # rejected rather than silently normalized on read.
        if self._canonical_bytes(payload) != encoded:
            raise WorkContextSnapshotError("context snapshot bytes are not canonical")
        _verify_snapshot_bytes(encoded, expected_digest)
        return payload

    def _snapshot_usage_unlocked(self) -> tuple[int, int]:
        count = 0
        total_bytes = 0
        try:
            with os.scandir(self.context_dir) as entries:
                for entry in entries:
                    if entry.name.startswith(".context-") and entry.name.endswith(".tmp"):
                        continue
                    if not entry.name.endswith(".json"):
                        continue
                    try:
                        info = entry.stat(follow_symlinks=False)
                    except OSError as exc:
                        raise WorkContextSnapshotError(
                            f"cannot inspect context snapshot entry: {exc}"
                        ) from exc
                    if stat.S_ISLNK(info.st_mode) or not stat.S_ISREG(info.st_mode):
                        raise WorkContextSnapshotError("unsafe context snapshot entry")
                    count += 1
                    total_bytes += int(info.st_size)
                    if count > self.max_snapshots or total_bytes > self.max_total_bytes:
                        break
        except OSError as exc:
            raise WorkContextSnapshotError(f"cannot enumerate context snapshots: {exc}") from exc
        return count, total_bytes

    def put_json(self, payload: dict[str, Any]) -> str:
        """Persist semantic context bytes and return ``wctx1.<context_sha256>``."""
        if not isinstance(payload, dict):
            raise ValueError("context snapshot payload must be an object")
        digest = self._context_commitment(payload)
        stable = self._stable_payload(payload)
        raw = self._canonical_bytes(stable)
        _verify_snapshot_bytes(raw, digest)
        target = self._path(digest)
        with self.graph_store.lock():
            _private_dir(self.context_dir)
            if target.exists() or target.is_symlink():
                existing = self._read_path_unlocked(target, digest)
                if self._canonical_bytes(existing) != raw:
                    raise WorkContextSnapshotError(
                        "context commitment maps to conflicting stable snapshot bytes"
                    )
                return self.token_for_commitment(digest)
            count, total_bytes = self._snapshot_usage_unlocked()
            if count >= self.max_snapshots:
                raise WorkContextSnapshotError(
                    "context snapshot store reached its bounded entry limit"
                )
            if total_bytes + len(raw) > self.max_total_bytes:
                raise WorkContextSnapshotError(
                    "context snapshot store reached its bounded byte limit"
                )
            fd, name = tempfile.mkstemp(
                prefix=".context-", suffix=".tmp", dir=self.context_dir
            )
            temp_path = Path(name)
            try:
                if os.name == "posix":
                    os.fchmod(fd, 0o600)
                with os.fdopen(fd, "wb", closefd=True) as handle:
                    handle.write(raw)
                    handle.flush()
                    os.fsync(handle.fileno())
                fd = -1
                if target.exists() or target.is_symlink():
                    existing = self._read_path_unlocked(target, digest)
                    if self._canonical_bytes(existing) != raw:
                        raise WorkContextSnapshotError(
                            "context commitment maps to conflicting stable snapshot bytes"
                        )
                else:
                    os.replace(temp_path, target)
                    if os.name == "posix":
                        target.chmod(0o600)
                    _fsync_dir(self.context_dir)
            finally:
                if fd >= 0:
                    os.close(fd)
                try:
                    temp_path.unlink(missing_ok=True)
                except OSError:
                    pass
        return self.token_for_commitment(digest)

    def get_json(self, token: str) -> dict[str, Any]:
        """Load only canonical bytes whose inner verified commitment matches."""
        digest = self._digest_from_token(token)
        with self.graph_store.lock():
            _private_dir(self.context_dir)
            return self._read_path_unlocked(self._path(digest), digest)


__all__ = [
    "CONTEXT_SNAPSHOT_TOKEN_PREFIX",
    "DEFAULT_MAX_CONTEXT_BYTES",
    "DEFAULT_MAX_SNAPSHOTS",
    "DEFAULT_MAX_TOTAL_BYTES",
    "WorkContextSnapshotError",
    "WorkContextSnapshotStore",
]

File: entroly-wasm/js/work_context_snapshot_store.js
'use strict';

// Host persistence mechanics only. The v1 verified-context commitment rules
// live in entroly-engine and are reached here through the generated WASM
// binding. This file owns repository isolation, bounds, atomic persistence and
// byte preservation; it does not define what makes a context snapshot valid.
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');
const { verifiedContextSnapshotVerifyBytes } = require('../pkg/entroly_wasm');
const { WorkGraphStateError, WorkGraphStore } = require('./work_graph_store');

const CONTEXT_SNAPSHOT_TOKEN_PREFIX = 'wctx1.';
const DEFAULT_MAX_CONTEXT_BYTES = 512 * 1024;
const DEFAULT_MAX_SNAPSHOTS = 8192;
const DEFAULT_MAX_TOTAL_BYTES = 256 * 1024 * 1024;
const DIGEST_RE = /^[0-9a-f]{64}$/;

class WorkContextSnapshotError extends WorkGraphStateError {}

function positiveInteger(value, name, minimum = 1) {
  if (!Number.isSafeInteger(value) || value < minimum) {
    throw new WorkContextSnapshotError(`${name} must be a safe integer >= ${minimum}`);
  }
  return value;
}

function secureDirectory(target) {
  fs.mkdirSync(target, { recursive: true, mode: 0o700 });
  const info = fs.lstatSync(target);
  if (info.isSymbolicLink() || !info.isDirectory()) {
    throw new WorkContextSnapshotError(`unsafe context snapshot directory: ${target}`);
  }
  if (process.platform !== 'win32') fs.chmodSync(target, 0o700);
}

function fsyncDirectory(target) {
  if (process.platform === 'win32') return;
  let fd;
  try {
    fd = fs.openSync(target, fs.constants.O_RDONLY | (fs.constants.O_DIRECTORY || 0));
    fs.fsyncSync(fd);
  } catch (_) {
    // Directory fsync is not supported by every filesystem/runtime. The Python
    // store has the same best-effort boundary; file data itself is fsynced.
  } finally {
    if (fd !== undefined) {
      try { fs.closeSync(fd); } catch (_) {}
    }
  }
}

function asBuffer(value) {
  if (Buffer.isBuffer(value)) return Buffer.from(value);
  if (value instanceof Uint8Array) return Buffer.from(value);
  if (typeof value === 'string') return Buffer.from(value, 'utf8');
  throw new WorkContextSnapshotError('context snapshot must be bytes or UTF-8 text');
}

/**
 * Verify exact snapshot bytes through the Rust semantic kernel.
 *
 * The byte stream is intentionally not normalized or reserialized before
 * verification. Python v1 commits its compact ensure-ASCII representation and
 * valid number lexemes such as `1.0` must survive a Python -> Node -> Python
 * handoff byte-for-byte.
 */
function verifyCanonicalSnapshotBytes(
  value,
  expectedCommitment,
  maxBytes = DEFAULT_MAX_CONTEXT_BYTES,
) {
  const bytes = asBuffer(value);
  positiveInteger(maxBytes, 'maxBytes', 1024);
  if (bytes.length > maxBytes) {
    throw new WorkContextSnapshotError(`context snapshot exceeds ${maxBytes} bytes`);
  }
  const expected = String(expectedCommitment || '');
  try {
    const commitment = verifiedContextSnapshotVerifyBytes(bytes, expected);
    // Parsing is presentation only and happens after Rust has accepted the exact
    // bytes. The parsed object is never reserialized for commitment verification.
    const payload = JSON.parse(bytes.toString('ascii'));
    return { payload, commitment, bytes };
  } catch (error) {
    if (error instanceof WorkContextSnapshotError) throw error;
    const detail = error && error.message ? error.message : String(error);
    throw new WorkContextSnapshotError(detail);
  }
}

class WorkContextSnapshotStore {
  constructor(graphStore, options = {}) {
    if (!(graphStore instanceof WorkGraphStore)) {
      throw new TypeError('graphStore must be a WorkGraphStore');
    }
    this.graphStore = graphStore;
    this.maxContextBytes = positiveInteger(
      options.maxContextBytes ?? DEFAULT_MAX_CONTEXT_BYTES,
      'maxContextBytes',
      1024,
    );
    this.maxSnapshots = positiveInteger(
      options.maxSnapshots ?? DEFAULT_MAX_SNAPSHOTS,
      'maxSnapshots',
    );
    if (this.maxSnapshots > 100000) {
      throw new WorkContextSnapshotError('maxSnapshots must not exceed 100000');
    }
    this.maxTotalBytes = positiveInteger(
      options.maxTotalBytes ?? DEFAULT_MAX_TOTAL_BYTES,
      'maxTotalBytes',
      this.maxContextBytes,
    );
    this.contextDir = path.join(graphStore.repoDir, 'context-snapshots');
    secureDirectory(this.contextDir);
  }

  static tokenForCommitment(commitment) {
    const digest = String(commitment || '');
    if (!DIGEST_RE.test(digest)) {
      throw new WorkContextSnapshotError('context commitment is not a sha256 digest');
    }
    return CONTEXT_SNAPSHOT_TOKEN_PREFIX + digest;
  }

  static digestFromToken(token) {
    const value = String(token || '');
    if (!value.startsWith(CONTEXT_SNAPSHOT_TOKEN_PREFIX)) {
      throw new WorkContextSnapshotError('unsupported context snapshot token');
    }
    const digest = value.slice(CONTEXT_SNAPSHOT_TOKEN_PREFIX.length);
    if (!DIGEST_RE.test(digest)) {
      throw new WorkContextSnapshotError('context snapshot token has an invalid digest');
    }
    return digest;
  }

  _snapshotPath(digest) {
    return path.join(this.contextDir, `${digest}.json`);
  }

  _readUnlocked(target, expectedDigest) {
    let info;
    try { info = fs.lstatSync(target); }
    catch (error) {
      if (error && error.code === 'ENOENT') {
        throw new WorkContextSnapshotError('context snapshot is unavailable');
      }
      throw new WorkContextSnapshotError(`cannot inspect context snapshot: ${error.message}`);
    }
    if (info.isSymbolicLink() || !info.isFile()) {
      throw new WorkContextSnapshotError('unsafe context snapshot path');
    }
    if (info.size > this.maxContextBytes) {
      throw new WorkContextSnapshotError('context snapshot exceeds its byte bound');
    }

    const flags = fs.constants.O_RDONLY | (fs.constants.O_NOFOLLOW || 0);
    let fd;
    try {
      fd = fs.openSync(target, flags);
      const current = fs.fstatSync(fd);
      if (!current.isFile() || current.size > this.maxContextBytes) {
        throw new WorkContextSnapshotError('unsafe or oversized context snapshot');
      }
      const bytes = fs.readFileSync(fd);
      const verified = verifyCanonicalSnapshotBytes(bytes, expectedDigest, this.maxContextBytes);
      return verified.bytes;
    } catch (error) {
      if (error instanceof WorkContextSnapshotError) throw error;
      throw new WorkContextSnapshotError(`cannot open context snapshot safely: ${error.message}`);
    } finally {
      if (fd !== undefined) {
        try { fs.closeSync(fd); } catch (_) {}
      }
    }
  }

  _usageUnlocked() {
    let count = 0;
    let totalBytes = 0;
    for (const name of fs.readdirSync(this.contextDir)) {
      if (name.startsWith('.context-') && name.endsWith('.tmp')) continue;
      if (!name.endsWith('.json')) continue;
      const target = path.join(this.contextDir, name);
      const info = fs.lstatSync(target);
      if (info.isSymbolicLink() || !info.isFile()) {
        throw new WorkContextSnapshotError('unsafe context snapshot entry');
      }
      count += 1;
      totalBytes += info.size;
      if (count > this.maxSnapshots || totalBytes > this.maxTotalBytes) break;
    }
    return { count, totalBytes };
  }

  putCanonicalBytes(value, expectedCommitment) {
    const expected = String(expectedCommitment || '');
    if (!DIGEST_RE.test(expected)) {
      throw new WorkContextSnapshotError('expectedCommitment must be a sha256 digest');
    }
    const verified = verifyCanonicalSnapshotBytes(value, expected, this.maxContextBytes);
    const token = WorkContextSnapshotStore.tokenForCommitment(expected);
    const target = this._snapshotPath(expected);

    return this.graphStore.withLock(() => {
      secureDirectory(this.contextDir);
      if (fs.existsSync(target)) {
        const existing = this._readUnlocked(target, expected);
        if (!existing.equals(verified.bytes)) {
          throw new WorkContextSnapshotError(
            'context commitment maps to conflicting stable snapshot bytes',
          );
        }
        return token;
      }
      const usage = this._usageUnlocked();
      if (usage.count >= this.maxSnapshots) {
        throw new WorkContextSnapshotError('context snapshot store reached its bounded entry limit');
      }
      if (usage.totalBytes + verified.bytes.length > this.maxTotalBytes) {
        throw new WorkContextSnapshotError('context snapshot store reached its bounded byte limit');
      }

      const temp = path.join(
        this.contextDir,
        `.context-${process.pid}-${crypto.randomUUID().replace(/-/g, '')}.tmp`,
      );
      let fd;
      try {
        fd = fs.openSync(temp, 'wx', 0o600);
        fs.writeFileSync(fd, verified.bytes);
        fs.fsyncSync(fd);
        fs.closeSync(fd);
        fd = undefined;
        if (fs.existsSync(target)) {
          const existing = this._readUnlocked(target, expected);
          if (!existing.equals(verified.bytes)) {
            throw new WorkContextSnapshotError(
              'context commitment maps to conflicting stable snapshot bytes',
            );
          }
        } else {
          fs.renameSync(temp, target);
          if (process.platform !== 'win32') fs.chmodSync(target, 0o600);
          fsyncDirectory(this.contextDir);
        }
      } finally {
        if (fd !== undefined) {
          try { fs.closeSync(fd); } catch (_) {}
        }
        try { fs.unlinkSync(temp); } catch (_) {}
      }
      return token;
    });
  }

  getCanonicalBytes(token) {
    const digest = WorkContextSnapshotStore.digestFromToken(token);
    return this.graphStore.withLock(() => {
      secureDirectory(this.contextDir);
      return this._readUnlocked(this._snapshotPath(digest), digest);
    });
  }

  getJSON(token) {
    return JSON.parse(this.getCanonicalBytes(token).toString('ascii'));
  }
}

module.exports = {
  CONTEXT_SNAPSHOT_TOKEN_PREFIX,
  DEFAULT_MAX_CONTEXT_BYTES,
  DEFAULT_MAX_SNAPSHOTS,
  DEFAULT_MAX_TOTAL_BYTES,
  WorkContextSnapshotError,
  WorkContextSnapshotStore,
  verifyCanonicalSnapshotBytes,
};
